common_types/src/bitset.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2022-2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! BitSet supports counting set/unset bits.
+
+#[derive(Debug, Default, Clone)]
+pub struct BitSet {
+    /// The bits are stored as bytes in the least significant bit order.
+    buffer: Vec<u8>,
+    /// The number of real bits in the `buffer`
+    num_bits: usize,
+}
+
+impl BitSet {
+    /// Initialize a unset [`BitSet`].
+    pub fn new(num_bits: usize) -> Self {
+        Self {
+            buffer: vec![0; Self::num_bytes(num_bits)],
+            num_bits,
+        }
+    }
+
+    #[inline]
+    pub fn num_bits(&self) -> usize {
+        self.num_bits
+    }
+
+    #[inline]
+    pub fn num_bytes(num_bits: usize) -> usize {
+        (num_bits + 7) >> 3
+    }
+
+    /// Initialize directly from a buffer.
+    ///
+    /// None will be returned if the buffer's length is not enough to cover the
+    /// bits of `num_bits`.
+    pub fn try_from_raw(buffer: Vec<u8>, num_bits: usize) -> Option<Self> {
+        if buffer.len() < Self::num_bytes(num_bits) {
+            None
+        } else {
+            Some(Self { buffer, num_bits })
+        }
+    }
+
+    /// Set the bit at the `index`.
+    ///
+    /// Return false if the index is outside the range.
+    pub fn set(&mut self, index: usize) -> bool {
+        if index >= self.num_bits {
+            return false;
+        }
+        let (byte_index, bit_index) = Self::compute_byte_bit_index(index);
+        self.buffer[byte_index] |= 1 << bit_index;
+        true
+    }
+
+    /// Tells whether the bit at the `index` is set.
+    pub fn is_set(&self, index: usize) -> Option<bool> {
+        if index >= self.num_bits {
+            return None;
+        }
+        let (byte_index, bit_index) = Self::compute_byte_bit_index(index);
+        let set = (self.buffer[byte_index] & (1 << bit_index)) != 0;
+        Some(set)
+    }
```

Comment:
Can another `unset` interface be provided?

Reply:
I guess there is no need to do so because after initialization all the bits are unset.